### ISSUE TITLE
Can't run a twine upload due to version problem

### ISSUE THREAD

**Issue by luriol** [pypi/warehouse#15611]
I'm trying to upload a distrubtion of a package to pypy using  python3 -m twine upload dist/*.16*

I get an error saying 
"
InvalidDistribution: Metadata is missing required fields: Name, Version.
"
If I unpack the whl file and look at the METADATA file I notice that an exact older version which uploaded correctly had the line

Metadata-Version: 2.1

When I currently try to build the package it won't upload (and also fails a twine check).  When I look at the METADATA from the newer version is says 

Metadata-Version: 2.3

If I simply change 2.3 to 2.1 I can rezip the whl package and it now passes the twine check and everything is fine.

How can I fix this?

**Comment by di** [comment]
Hi, are you using the [latest version of `twine`](https://pypi.org/project/twine/5.0.0/)?

**Comment by stecohen** [comment]
I have the same issue

**Comment by di** [comment]
Are you using the [latest version of `twine`](https://pypi.org/project/twine/5.0.0/)?

**Comment by stecohen** [comment]
Yes, as per my initial post

**Comment by stecohen** [comment]
I had to downgrade python from 3.11.7 to 3.10.10 to fix this issue

**Comment by umarbutler** [comment]
+1 Same issue here. I can confirm that the exact same command I used to upload a package on the 15th (`py -m twine upload --repository pypi dist/*`) raised this issue today when trying to upload a new version of the same package. I upgraded to the latest version of `twine` and it did not help.

**Comment by LightOfHeaven1994** [comment]
@umarbutler thanks a lot, bumping `pkginfo` helped us to resolve the issue :heart:

**Comment by LtStarbuck1030** [comment]
> On 15 March, I updated a package using `py -m twine upload --repository pypi dist/*` and experienced no issues. Today, I built a new package, seemingly having not upgraded any of my build tools (although perhaps it is possible I did?) and experienced this issue: `InvalidDistribution: Metadata is missing required fields: Name, Version.`. This persisted even after upgrading my `twine` and trying Python 11 instead of 12. Only after upgrading `pkginfo` was the issue fixed. A higher version of `pkginfo` should be a requirement to ensure that this doesn't happen to others. At the very least, `twine`'s error message should mentioned that upgrading `pkginfo` will probably solve the issue.
> 
> Related issues I have located are as follows:
> 
>     * [Check command doesn't recognize metadata 2.3 pypa/twine#1059](https://github.com/pypa/twine/issues/1059)
> 
>     * [Can't run a twine upload due to version problem #15611](https://github.com/pypi/warehouse/issues/15611)
> 
>     * [Twine should require pkginfo > 1.10 pypa/twine#1070](https://github.com/pypa/twine/issues/1070)
> 
>     * [Successfully build *.whl and *.tar.gz, but cannot upload them through twine pypa/packaging-problems#735](https://github.com/pypa/packaging-problems/issues/735)

This worked perfectly. Ashamed of how long it took me to stumble on this response. Thank you <3

**Comment by pultarmi** [comment]
This is still a problem with twine==5.1.1, pkginfo==1.11.1, hatchling==1.25.0, python=3.12.4 on Ubuntu 24. I tried multiple versions of the packages and Python, it's always the same error message

```
ERROR    InvalidDistribution: Metadata is missing required fields: Name, Version.                                                                                                 
         Make sure the distribution includes the files where those fields are specified, and is using a supported Metadata-Version: 1.0, 1.1, 1.2, 2.0, 2.1, 2.2.
```

There is `Metadata-Version: 2.3` isnide PKG_INFO in dist folder, so I guess that hatchling uses a newer metadata format than what twine supports.

**Comment by dmitrycrivoy** [comment]
> This is still a problem with twine==5.1.1, pkginfo==1.11.1, hatchling==1.25.0, python=3.12.4 on Ubuntu 24. I tried multiple versions of the packages and Python, it's always the same error message
> 
> ```
> ERROR    InvalidDistribution: Metadata is missing required fields: Name, Version.                                                                                                 
>          Make sure the distribution includes the files where those fields are specified, and is using a supported Metadata-Version: 1.0, 1.1, 1.2, 2.0, 2.1, 2.2.
> ```
> 
> There is `Metadata-Version: 2.3` isnide PKG_INFO in dist folder, so I guess that hatchling uses a newer metadata format than what twine supports.
> 
> As I see it, the only working workaround is to modify the header in PKG_INFO manually.

Interesting thing.
I upgraded **twine** up to **5.1.1**, then I tried to upgrade **pkginfo** up to latest version (currently **1.11.1**) and got an error `twine 5.1.1 requires pkginfo<1.11, but you have pkginfo 1.11.1 which is incompatible.`, so I run this command and it worked: 
`python3.11 -m pip install --upgrade pkginfo==1.10` 

Make sure to use your specific python version if it's applicable to you.

**Comment by progeroffline** [comment]
After update package `pkginfo` the problem went away.
```bash
python3 -m pip install --upgrade pkginfo
```

**Comment by BrainAnnex** [comment]
**SHORT VERSION** : updating `twine` does NOT resolve the problem.  Fortunately, updating `pkginfo` fixes everything :)  

----
I slammed against this same issue today.  I did everything IDENTICALLY to my usual protocol that had worked just fine until 1 month ago, to generate the distribution package, namely running  `python3 -m build `  (in the top directory of my project), 
which created the usual `tar.gz` and `py3-none-any.whl` files.

So far so good - but then attempting to run `python3 -m twine upload dist/* ` (which always worked before) suddenly started barking:

'''
Uploading distributions to https://upload.pypi.org/legacy/
ERROR    InvalidDistribution: Metadata is missing required fields: Name,        
         Version.                                                               
         Make sure the distribution includes the files where those fields are   
         specified, and is using a supported Metadata-Version: 1.0, 1.1, 1.2,   
         2.0, 2.1, 2.2, 2.3. `
'''
Very puzzling, before I hadn't changed anything in the OS (a Linux Mint 21.2 laptop dedicated to PyPI uploads).

When I double-clicked on the `tar.gz` and `py3-none-any.whl` files, drilling down until I found `PKG-INFO` files, and opened them, I saw **`Metadata-Version: 2.4`** .
By contrast, to my astonishment, I saw `Metadata-Version: 2.3` in the old files that I still had around from last month! 
I have no idea how this happened!!  Could it possibly have to do with the fact that in `.toml` file for my project I switched from requires-python = ">=3.8"  to 3.9 ???  Puzzled!

At any rate, I found out which version of `twine` I had:  `twine --version` showed 5.1.1
The `twine` software is a local install, as showed by `which twine` (which gave a `/.local/bin` directory)... and so I updated it, to the current latest version 6.0.1 by means of : `python3 -m pip install --upgrade twine`

As others on this thread have pointed out, the problem did NOT go away by upgrading `twine`.

As suggested on this thread, I then investigated `pkginfo` :  I had version 1.10.0 installed locally.  I upgraded it with:

`python3 -m pip install --upgrade pkginfo`

and, MIRACLE, `twine` (which is now version 6.0.1) works :) 

PS:  puzzled by the word "legacy" in `Uploading distributions to https://upload.pypi.org/legacy/` , but that seem a benign quirk...

**Comment by jhcole** [comment]
Found this thread working through the error in https://github.com/pypi/warehouse/issues/15611#issuecomment-2198363695

After updating Twine to 6.1.0, and pkginfo to 1.12.0 (latest at time of writing) I get this error.
```
ERROR    InvalidDistribution: Invalid distribution metadata: unrecognized or malformed field 'license-file'
```
Downgrading Twine to 6.0.1 worked.

**Comment by di** [comment]
@jhcole Is it possible your metadata is actually malformed? Can you show us what you're trying to build or provide the distribution you're trying to upload?

**Comment by ariebovenberg** [comment]
@di I ran into the same issue. Downgrading to `twine==6.0.1` fixed it. I'm not sure if it's malformed—I'm not that familiar with the format. It's generated by `maturin`. Here is the `METADATA`:

```
Metadata-Version: 2.4
Name: whenever
Version: 0.6.17
Classifier: Development Status :: 4 - Beta
Classifier: Intended Audience :: Developers
Classifier: License :: OSI Approved :: MIT License
Classifier: Operating System :: MacOS
Classifier: Operating System :: Microsoft :: Windows
Classifier: Operating System :: POSIX :: Linux
Classifier: Programming Language :: Python :: 3
Classifier: Programming Language :: Python :: 3.9
Classifier: Programming Language :: Python :: 3.10
Classifier: Programming Language :: Python :: 3.11
Classifier: Programming Language :: Python :: 3.12
Classifier: Programming Language :: Python :: 3.13
Classifier: Programming Language :: Python :: Implementation :: CPython
Classifier: Programming Language :: Python :: Implementation :: PyPy
Classifier: Programming Language :: Python
Classifier: Programming Language :: Rust
Classifier: Typing :: Typed
Requires-Dist: tzdata >=2020.1 ; sys_platform == 'win32'
License-File: LICENSE
License-File: LICENSE-THIRD-PARTY
Summary: Modern datetime library for Python
Keywords: datetime,typesafe,rust,date,time,timezone,utc,aware,zoneinfo,tzdata,tzdb
Author-email: Arie Bovenberg <a.c.bovenberg@gmail.com>
Maintainer-email: Arie Bovenberg <a.c.bovenberg@gmail.com>
License: MIT
Requires-Python: >=3.9
Description-Content-Type: text/markdown; charset=UTF-8; variant=GFM
Project-URL: Documentation, https://whenever.readthedocs.io
Project-URL: Repository, https://github.com/ariebovenberg/whenever
Project-URL: Issues, https://github.com/ariebovenberg/whenever/issues
Project-URL: Changelog, https://github.com/ariebovenberg/whenever/blob/master/CHANGELOG.rst

[readme omitted for brevity]
```

**Comment by di** [comment]
@jhcole @ariebovenberg Looks like this is https://github.com/pypa/twine/issues/1216

**Comment by pacman82** [comment]
Can confirm, have the same symptom for the build of <https://github.com/pacman82/arrow-odbc-py>, stumbled on this thread debugging CI. Downgrading to twine 6.0.1 helped me also, but I still do not know what's going on here.

Best, Markus

**Comment by attack68** [comment]
Same issue. Also using Maturin. Just downgraded to 6.0.1

**Comment by saifeiLee** [comment]
> On 15 March, I updated a package using `py -m twine upload --repository pypi dist/*` and experienced no issues. Today, I built a new package, seemingly having not upgraded any of my build tools (although perhaps it is possible I did?) and experienced this issue: `InvalidDistribution: Metadata is missing required fields: Name, Version.`. This persisted even after upgrading my `twine` and trying Python 11 instead of 12. Only after upgrading `pkginfo` was the issue fixed. A higher version of `pkginfo` should be a requirement to ensure that this doesn't happen to others. At the very least, `twine`'s error message should mention that upgrading `pkginfo` will probably solve the issue.
> 
> Related issues I have located are as follows:
> 
> * [Check command doesn't recognize metadata 2.3 pypa/twine#1059](https://github.com/pypa/twine/issues/1059)
> * [Can't run a twine upload due to version problem #15611](https://github.com/pypi/warehouse/issues/15611)
> * [Twine should require pkginfo > 1.10 pypa/twine#1070](https://github.com/pypa/twine/issues/1070)
> * [Successfully build *.whl and *.tar.gz, but cannot upload them through twine pypa/packaging-problems#735](https://github.com/pypa/packaging-problems/issues/735)

Thanks a lot. It works for me.

**Comment by mhearne-usgs** [comment]
Does anyone have a workaround for this issue? Upgrading to the most recent version of twine results in this error: "pypi InvalidDistribution: Invalid distribution metadata: unrecognized or malformed field 'license-file'". Pinning to twine 6.0.1 gives me the error about the wrong metadata versions: "InvalidDistribution: Make sure the distribution is using a supported Metadata-Version: 1.0, 1.1, 1.2, 2.0, 2.1, 2.2, 2.3" I ran build on my laptop and I see that the Metadata-Version is set to 2.4. I'm doing my build in a CI system, so I can't manually change this. I'm about to embark on adding a script to my repo to unzip the wheel, fix the metadata version, then rezip-the wheel before calling twine. If somebody has a better method, please stop me!

**Comment by abm11** [comment]
> Ok, I solved my problem by pinning to twine==6.0.1 and setuptools==75.6.0. I don't like having pins to specific older versions of packages in my pyproject.toml files so hopefully I learn of a better approach.

Worked for me. I've been going up the wall trying to fix this. Thank you.

**Comment by princenyeche** [comment]
Upgraded twine version 6.1.0 - fixed it for me on Python 3.10

**Comment by matkoniecz** [comment]
switched source from apt package to pipx, this allowed to install version without this bug

**Comment by harryli0088** [comment]
In case anyone is encountering this with setuptools, I had to pin my version to setuptools==76.1.0 to publish to PyPI. It looks like setuptools versions 77 and higher output Metadata version 2.4, which is too high for PyPI

**Comment by di** [comment]
Hi, PyPI has supported Metadata 2.4 since #16949 was merged in Nov 2024. It's likely that you're having a different issue than OP and that pinning setuptools is not the right solution, I would recommend opening a new issue at https://github.com/pypa/setuptools/issues with more details.

Since the original issue is resolved, I'll lock this thread. Anyone who thinks they are experiencing this should open a new issue.
